pkg/operators/oci/oci.go

```diff
@@ -0,0 +1,275 @@
+// were the gadget
+type ociHandler struct{}
+
+func (o *ociHandler) ParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		// Hardcoded for now
+		{
+			Key:          authfileParam,
+			Title:        "Auth file",
+			Description:  "Path of the authentication file. This overrides the REGISTRY_AUTH_FILE environment variable",
+			DefaultValue: oci.DefaultAuthFile,
+			TypeHint:     params.TypeString,
+		},
+		{
+			Key:          validateMetadataParam,
+			Title:        "Validate metadata",
+			Description:  "Validate the gadget metadata before running the gadget",
+			DefaultValue: "true",
+			TypeHint:     params.TypeBool,
+		},
+		{
+			Key:          insecureParam,
+			Title:        "Insecure connection",
+			Description:  "Allow connections to HTTP only registries",
+			DefaultValue: "false",
+			TypeHint:     params.TypeBool,
+		},
+		{
+			Key:          pullParam,
+			Title:        "Pull policy",
+			Description:  "Specify when the gadget image should be pulled",
+			DefaultValue: oci.PullImageMissing,
+			PossibleValues: []string{
+				oci.PullImageAlways,
+				oci.PullImageMissing,
+				oci.PullImageNever,
+			},
+			TypeHint: params.TypeString,
+		},
+		{
+			Key:         pullSecret,
+			Title:       "Pull secret",
+			Description: "Secret to use when pulling the gadget image",
+			TypeHint:    params.TypeString,
+		},
+	}
+}
+
+func (o *ociHandler) Instantiate(gadgetCtx operators.GadgetContext, params *params.Params) (*OciHandlerInstance, error) {
+	if len(gadgetCtx.ImageName()) == 0 {
+		return nil, fmt.Errorf("imageName empty")
+	}
+
+	secretBytes := []byte{} // TODO
+
+	authOpts := &oci.AuthOptions{
+		AuthFile:    params.Get(authfileParam).AsString(),
+		SecretBytes: secretBytes,
+		Insecure:    params.Get(insecureParam).AsBool(),
+	}
+
+	// Make sure the image is available, either through pulling or by just accessing a local copy
+	// TODO: add security constraints (e.g. don't allow pulling - add GlobalParams for that)
+	err := oci.EnsureImage(gadgetCtx.Context(), gadgetCtx.ImageName(), authOpts, params.Get(pullParam).AsString())
+	if err != nil {
+		return nil, fmt.Errorf("insuring image: %w", err)
+	}
+
+	manifest, err := oci.GetManifestForHost(gadgetCtx.Context(), gadgetCtx.ImageName())
+	if err != nil {
+		return nil, fmt.Errorf("getting manifest: %w", err)
+	}
+
+	logger := gadgetCtx.Logger()
+	logger.Debugf("ArtifactType: %s", manifest.ArtifactType)
+
+	// metadata := &types.GadgetMetadata{}
+	r, err := oci.GetContentFromDescriptor(gadgetCtx.Context(), manifest.Config)
+	if err != nil {
+		return nil, fmt.Errorf("getting metadata: %w", err)
+	}
+	metadata, _ := io.ReadAll(r)
+	r.Close()
+
+	// Store metadata for serialization
+	gadgetCtx.SetMetadata(metadata)
+
+	viper := viper.New()
+	viper.SetConfigType("yaml")
+	err = viper.ReadConfig(bytes.NewReader(metadata))
+
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling metadata: %w", err)
+	}
```

Comment:
Yes, it's only used for parsing and it's not using our global instance (the `viper.New()` takes care of that). I'm not 100% sure about this part as it feels a bit hacky. I wanted the underlying yaml reader to not be too strict, so that it would not drop fields that we don't specify in Go structs.
For example, wasm could also be using an interface to get certain sections (or fields directly) of the metadata, that are specific to wasm. Or third party operators could access them without having to extend internal data structures in the core parts of IG.

Reply:
Mmmmm yeah, I found it a bit hacky. I initially thought we could store the underlying yaml and then whoever needs to use it will marshall it into a struct that contains only the field they need. WDYT?

Reply:
Mmmmm something like [this](https://go.dev/play/p/rruJNUGvvfR)?

Reply:
Oh, btw, I don't know if it can be done with yaml too 😅 . That's what you meant?

Reply:
> Mmmmm something like [this](https://go.dev/play/p/rruJNUGvvfR)?

That's just normal JSON unmarshaling, right? The point is, what if you're only interested in MetaData.Tracers[0].MyExtension (where MetaData is a struct, Tracers an array of Structs, and MyExtension your self-defined third-party definition). Using viper, you can exactly do that: get `MetaData.Tracers[0].MyExtension` and then unmarshal that into your struct (it's essentially storing everything as `map[string]any` internally). If we unmarshal into hardcoded structs, we dismiss any additional fields that we don't know about in code.

In JSON you have `json.RawMessage` that can do partially what I mean (but also not fully) - the yaml packages don't have something like that.

Reply:
Let me think about it. In the meanwhile, could you please give me an example of what MyExtension would be?

Reply:
> Let me think about it. In the meanwhile, could you please give me an example of what MyExtension would be?

Could be

```yaml
myWASMOptions:
  foo: 5
```

for example, which could be forwarded to WASM.

Or could be a configuration tailored to a third-party operator somehow.

But maybe that's too far ahead in the future and isn't worth implementing that flexible now.

Reply:
I'd also prefer to keep this simple for now. We're not sure about how to handle it and we don't have an immediate need, so I think just storing the metadata should be good enough for now.

Reply:
I also agree (hence the comment) 😄

My worry is that all entities using parts of the metadata that way would need to know/depend on the entire metadata structure, which doesn't feel modular.

I'll remove it for now and will try to find a better solution later.

Reply:
Update: will be addresses in a follow-up PR.